src/xwayland.rs

```diff
@@ -136,6 +136,104 @@ impl State {
+                            std::iter::once(active.clone()).chain(
+                                mapped
+                                    .is_stack()
+                                    .then(move || {
+                                        mapped
+                                            .windows()
+                                            .map(|(s, _)| s)
+                                            .filter(move |s| s != &active)
+                                    })
+                                    .into_iter()
+                                    .flatten(),
+                            )
+                        })
+                        .chain(
+                            set.workspaces
+                                .iter()
+                                .enumerate()
+                                .filter(|(i, _)| *i == set.active)
+                                .chain(
+                                    set.workspaces
+                                        .iter()
+                                        .enumerate()
+                                        .filter(|(i, _)| *i != set.active),
+                                )
+                                .flat_map(|(_, workspace)| {
+                                    workspace.get_fullscreen().cloned().into_iter().chain(
+                                        workspace.mapped().flat_map(|mapped| {
+                                            let active = mapped.active_window();
+                                            std::iter::once(active.clone()).chain(
+                                                mapped
+                                                    .is_stack()
+                                                    .then(move || {
+                                                        mapped
+                                                            .windows()
+                                                            .map(|(s, _)| s)
+                                                            .filter(move |s| s != &active)
+                                                    })
+                                                    .into_iter()
+                                                    .flatten(),
+                                            )
+                                        }),
+                                    )
+                                }),
+                        )
+                })
+                .collect::<Vec<_>>();
+
+            // we don't include the popup elements, which contain the OR windows, because we are not supposed to restack them.
+            // Which is also why we match upwards, to not disturb elements at the top.
+            //
+            // But this also means we need to match across all outputs and workspaces at once, to be sure nothing that shouldn't be on top of us is.
+            if let Err(err) = xwm.update_stacking_order_upwards(order.iter().rev()) {
```

Comment:
Is `update_stacking_order_upwards()` with `rev()` different from `update_stacking_order_downwards()`?

It could be good to pass the iterator here without `collect, but I guess that wouldn't be possible with `update_stacking_order_*` taking `impl Iterator<Item = &'a W>`. (Maybe it could use a `Borrow` generic.)

Reply:
(I guess it is different, based on the "Which is also why we match upwards, to not disturb elements at the top" comment.)

Reply:
Yes, sadly it is different. Essentially we don't want to touch the top-most elements of the stacking order, which `_downwards` could potentially do.